### DISCUSSION TITLE
Does anyone have a way to Globally Update Runtime Config OR App Config at runtime?

### DISCUSSION THREAD

**Discussion by Clive-Ward-Cron** [nuxt/nuxt#22098] [Questions] [answered]
## The Reason
I have some configuration data that I need to get from an external API, but its an expensive call and I don't want to make the request more times than necessary. To solve this issue I've made a module that makes the request at build time, and I've been able to easily add the data to the RuntimeConfig or the AppConfig. This makes it available through out the app to all users without having to fetch the data on every page load or something.

## The Problem
The issue I'm running into is that this data may change at any time, and the application should preferably reflect some of these changes without the need to rebuild the entire app. I'd also like to add some fault tolerance because the call for the data could potentially fail if the external API is down or etc. 

## What I've Tried
I tried adding the data to RuntimeConfig but the only way I've seen to update that at runtime was using env variables and I can't place all of this information in env variables.

So instead I'm placing this data in AppConfig on build if its successful. If its not successful, I try to fetch the data later on at runtime by using a middleware that checks if the data is available from `useAppConfig` and fetches the data if it isn't found. Then I update the AppConfig using `updateAppConfig({myProperty})`. 

So that works, ***BUT*** it only works for that one request and the updated AppConfig isn't persisted globally. If I refresh the page or open the app in another browser, the updated AppConfig is no longer there and the request has to be made again. So I end up having to make the expensive request on every page load anyway.

## Questions:
+ Does anyone know of a work around for what I'm trying to do here?
+ Should I be doing this another way entirely?

**Comment by jwhazel** [comment, marked as answer]
> Should I be doing this another way entirely?

This sounds like a candidate for creating a config provider API proxy and using stale-while-revalidate in Nitro so that it caches the data.

1.) create the following file `server/api/config.get.js` (more reading: https://nuxt.com/docs/guide/directory-structure/server)
2.) put your logic for querying the API there and returning the results you specifically want in your app
3.) set up your nuxt config file so that it caches this route with swr, set it to whatever TTL you want (more reading: https://nitro.unjs.io/guide/cache#route-rules)
4.) use `const {data:config} = await useFetch('/api/config')` in your page/component where needed.

If for example you set your TTL to 60 minutes, your origin API would never get hit more than 24 times a day no matter if you had 100k page loads. This also gives you some control over what happens on a failed origin request (serve default JSON?)

Your other option is to do it entirely client side as a reactive store (more reading: https://nuxt.com/docs/getting-started/state-management). In this case you would create a composable that calls the origin API and stores the data with `useState()` (or just use Pinia: https://pinia.vuejs.org/introduction.html). Then you would call that store across any page/component that needed it and it would persist across the whole app. The mild downside is that it can still hammer your origin API depending on how many users you have (but not how many pages are loaded) and you would also have to manually implement the logic determining when to refresh the data. It also wouldn't persist if users opened the app in a new tab/browser or refreshed the page.

**Reply by Clive-Ward-Cron** [reply]
> > Should I be doing this another way entirely?
> 
> This sounds like a candidate for creating a config provider API proxy and using stale-while-revalidate in Nitro so that it caches the data.
> 
> 1.) create the following file `server/api/config.get.js` (more reading: https://nuxt.com/docs/guide/directory-structure/server)
> 2.) put your logic for querying the API there and returning the results you specifically want in your app
> 3.) set up your nuxt config file so that it caches this route with swr, set it to whatever TTL you want (more reading: https://nitro.unjs.io/guide/cache#route-rules)
> 4.) use `const {data:config} = await useFetch('/api/config')` in your page/component where needed.
> 
> If for example you set your TTL to 60 minutes, your origin API would never get hit more than 24 times a day no matter if you had 100k page loads. This also gives you some control over what happens on a failed origin request (serve default JSON?)
> 
> Your other option is to do it entirely client side as a reactive store (more reading: https://nuxt.com/docs/getting-started/state-management). In this case you would create a composable that calls the origin API and stores the data with `useState()` (or just use Pinia: https://pinia.vuejs.org/introduction.html). Then you would call that store across any page/component that needed it and it would persist across the whole app. The mild downside is that it can still hammer your origin API depending on how many users you have (but not how many pages are loaded) and you would also have to manually implement the logic determining when to refresh the data. It also wouldn't persist if users opened the app in a new tab/browser or refreshed the page.
> 
> 

I think that first option is probably what I'm going to need. I'll give that a try and let you know if it works for this use case.

**Reply by Clive-Ward-Cron** [reply]
@jwhazel thanks for sending me down the right path for this one.

I went with the SWR/TTL implementation but with a slightly different approach. 

I'm building this out in a module that I'd like to make available to other internal projects, so I didn't want to hard code those routes into the nuxt.config `routeRules` property in case I need to change the route name for all projects at some point. I also needed to be able to have a way to invalidate the cache in case I need to immediately refresh the data after a config change.

Here was my final solution:

1. In my local modules `runtime > handlers` directory I added a `utils` directory with an exported Nitro caching function.
2. The caching utility uses Nitro's `cachedFunction` to wrap my external call.
3. My `cachedFunction` accepts an `invalidate` arg that is defaulted to `false`.
4. I then pass `CacheOptions` to my `cachedFunction` and set the `swr` and `maxAge` properties as necessary along with a function for `shouldInvalidate` that resolves based on the arg passed to the main function.
5. Now instead of making the external call directly in my handler, I use the `cachedFunction` and I can optionally invalidate it based on some logic.

```ts
export const cachedConfig = cachedFunction(async (groupConfig: string, invalidate = false) => {
  const URL = `/Config/${group}/`
  const { data } = await useCustomFetch<CustomResponse<CustomConfig>>(URL)
  return data
}, {
  maxAge: 60 * 15, // Cache Results for 15 minutes
  swr: true, // Stale-While-Revalidate
  name: 'customConfig',
  getKey: (configGroup: string) => `${configGroup}-config`,
  // Invalidate the Cache as necessary based on passed arg
  //? args to the main function are being spread here, so they all have to be here
  shouldInvalidateCache: (configGroup: string, invalidate: boolean) => {
    if (invalidate) console.log('Invalidating Config Cache')
    return invalidate
  }
})
```
